src/herring/node.hpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cstdint>
+
+namespace herring {
+  /* Summary of Types
+   * ----------------
+   *  value_t (float or double): The value used for testing a node condition or
+   *    for providing the output of leaves.
+   *  feature_index_t (std::uint16_t or std::uint32_t): Index indicating which
+   *    feature this conditional applies to
+   *  offset_t (std::uint16_t or std::uint32_t): Offset between this node and
+   *    its distant child. For small trees, using a smaller type can reduce the
+   *    padded size of the node to as few as 8 bytes.
+   *  output_index_t (typically std::uint32_t): If leaf output values cannot be stored
+   *    in the same memory as test condition values, this index provides a
+   *    lookup location for output values stored in the tree.
+   */
+  template<typename value_t, typename feature_index_t, typename offset_t, typename output_index_t>
+  struct simple_node {
+    using value_type = value_t;  // float or double
+    using index_type = feature_index_t;
+    using offset_type = offset_t;
+    using output_index_type = output_index_t;
+    // Cannot use std::variant here because it takes up 4 additional bytes when
+    // value_type is float
+    union value_or_index {
+      value_type value;
+      output_index_type index;
+    };
+    value_or_index value;  // 4 bytes for float
+    offset_type distant_offset;  // 2 bytes for depth < 16 or small trees; 4 otherwise
+    index_type feature; // 1-4 bytes, depending on number of features
+  };
```

Comment:
Given the bit-packing here, it could be a good idea to add `static_assert` to ensure that `sizeof(simple_node)` within an acceptable range.

Reply:
In general, it's not a terrible issue if the size goes above what we expect, and as we support additional types, it may indeed get larger or smaller than is outlined here. The point is primarily to ensure that the node is as small as possible given the types that are used to instantiate the template rather than trying to meet any specific target for e.g. number of nodes in a cache line. Given that, I don't think a `static_assert` tells us anything useful beyond "Hey! You're using a larger type than you used to use," which should be obvious when we update the available types in the `std::variant` declaration.